client/client.go

```diff
@@ -128,15 +151,74 @@ func (t userAgentTransport) RoundTrip(req *http.Request) (*http.Response, error)
 	newReq.Header["Authorization"] = []string{fmt.Sprintf("Bearer %s", t.client.BearerAuthorizer.tokenProvider.OAuthToken())}
 
 	var retries int
-	for retries = 0; retries < throttlingAdditionalRetryCount; retries++ {
+	for retries = 0; retries < retryMax; retries++ {
 		response, err := t.base.RoundTrip(&newReq)
-		if err == nil && response.StatusCode == 429 {
-			// We hit throttling, retry to hopefully hit another ARM instance.
-			continue
+
+		shouldRetry, rerr := retryPolicy(response, err)
+		if !shouldRetry {
+			return response, err
 		}
+		wait := backoff(retryWaitMin, retryWaitMax, retries, response)
+
+		l := log.G(context.TODO()).WithField("URL", newReq.URL)
+		l.Infof("would retry on error %s, waiting for %s", rerr, wait)
 
-		return response, err
+		time.Sleep(wait)
 	}
 
 	return t.base.RoundTrip(&newReq)
 }
+
+func retryPolicy(resp *http.Response, err error) (bool, error) {
+	if err != nil {
+		if v, ok := err.(*url.Error); ok {
+			// Don't retry if the error was due to too many redirects.
+			if redirectsErrorRe.MatchString(v.Error()) {
+				return false, v
+			}
+
+			// Don't retry if the error was due to an invalid protocol scheme.
+			if schemeErrorRe.MatchString(v.Error()) {
+				return false, v
+			}
+
+			// Don't retry if the error was due to TLS cert verification failure.
+			if _, ok := v.Err.(x509.UnknownAuthorityError); ok {
+				return false, v
+			}
+		}
+
+		// The error is likely recoverable so retry. this includes
+		// conection closed, connection failure, timeout, request canceled, etc.
+		return true, nil
+	}
+
+	// refer to https://docs.microsoft.com/en-us/azure/architecture/best-practices/retry-service-specific#general-rest-and-retry-guidelines
+	if resp.StatusCode == http.StatusTooManyRequests ||
```

Comment:
define a const for those http status code list?

Reply:
do you mean adding a const that list all retrayable status code? if yes, I guess I can use https://github.com/Azure/go-autorest/blob/master/autorest/client.go#L48 directly.

Reply:
yes. but we may have different retry status codes here (e.g. cloud provider is using a different list).

Reply:
added